Mathlib/AlgebraicGeometry/Morphisms/Immersion.lean

```diff
@@ -0,0 +1,167 @@
+  have : IsPreimmersion (f.liftCoborder ≫ f.coborderRange.ι) := by
+    simp only [Scheme.Hom.liftCoborder_ι]; infer_instance
+  have : IsPreimmersion f.liftCoborder := .of_comp f.liftCoborder f.coborderRange.ι
+  refine .of_isPreimmersion _ ?_
+  convert isClosed_preimage_val_coborder
+  apply Set.image_injective.mpr f.coborderRange.ι.embedding.inj
+  rw [← Set.range_comp, ← TopCat.coe_comp, ← Scheme.comp_base, f.liftCoborder_ι]
+  exact (Set.image_preimage_eq_of_subset (by simpa using subset_coborder)).symm
+
+instance [IsImmersion f] : IsDominant f.coborderRange.ι := by
+  rw [isDominant_iff, DenseRange, Scheme.Opens.range_ι]
+  exact dense_coborder
+
+lemma isImmersion_eq_inf : @IsImmersion = (@IsPreimmersion ⊓
+    topologically fun {X Y} _ _ f ↦ IsLocallyClosed (Set.range f) : MorphismProperty Scheme) := by
+  ext; exact isImmersion_iff _
+
+namespace IsImmersion
+
+instance : IsLocalAtTarget @IsImmersion := by
+  suffices IsLocalAtTarget (topologically fun {X Y} _ _ f ↦ IsLocallyClosed (Set.range f)) from
+    isImmersion_eq_inf ▸ inferInstance
+  apply (config := { allowSynthFailures := true }) topologically_isLocalAtTarget'
+  · refine { precomp := ?_, postcomp := ?_ }
+    · intro X Y Z i hi f hf
+      replace hi : IsIso i := hi
+      show IsLocallyClosed _
+      simpa only [Scheme.comp_coeBase, TopCat.coe_comp, Set.range_comp,
+        Set.range_iff_surjective.mpr i.surjective, Set.image_univ]
+    · intro X Y Z i hi f hf
+      replace hi : IsIso i := hi
+      show IsLocallyClosed _
+      simp only [Scheme.comp_coeBase, TopCat.coe_comp, Set.range_comp]
+      refine hf.image i.homeomorph.inducing ?_
+      rw [Set.range_iff_surjective.mpr i.surjective]
+      exact isOpen_univ.isLocallyClosed
+  · simp_rw [Set.range_restrictPreimage]
+    exact fun _ _ _ e _ ↦ isLocallyClosed_iff_coe_preimage_of_iSup_eq_top e _
+
+instance (priority := 900) {X Y : Scheme} (f : X ⟶ Y) [IsOpenImmersion f] : IsImmersion f where
+  isLocallyClosed_range := f.isOpenEmbedding.2.isLocallyClosed
+
+instance (priority := 900) {X Y : Scheme} (f : X ⟶ Y) [IsClosedImmersion f] : IsImmersion f where
+  isLocallyClosed_range := f.isClosedEmbedding.2.isLocallyClosed
+
+instance : MorphismProperty.IsMultiplicative @IsImmersion where
+  id_mem _ := inferInstance
+  comp_mem {X Y Z} f g hf hg := by
+    refine { __ := inferInstanceAs (IsPreimmersion (f ≫ g)), isLocallyClosed_range := ?_ }
+    simp only [Scheme.comp_coeBase, TopCat.coe_comp, Set.range_comp]
+    exact f.isLocallyClosed_range.image g.embedding.toInducing g.isLocallyClosed_range
+
+instance comp {X Y Z : Scheme} (f : X ⟶ Y) (g : Y ⟶ Z) [IsImmersion f]
+    [IsImmersion g] : IsImmersion (f ≫ g) :=
+  MorphismProperty.IsStableUnderComposition.comp_mem f g inferInstance inferInstance
+
+variable {f} in
+/--
+A morphism is a (locally-closed) immersion if and only if it can be factored into
+a closed immersion followed by a (dominant) open immersion.
```

Comment:
In this lemma, we do not use this condition:
```suggestion
a closed immersion followed by an open immersion.
```

Reply:
Does it mean `locally-closed => dominant`?